File: frontend/src/types/todo.ts
export type TodoPriority = 'low' | 'medium' | 'high'

export type TodoStatus = 'todo' | 'in_progress' | 'done'

export type TodoItem = {
  id: string
  title: string
  description: string | null
  dueDate: string | null
  priority: TodoPriority
  status: TodoStatus
  tags: string[]
  insertedAt: string | null
  updatedAt: string | null
}

export type CreateTodoInput = {
  title: string
  priority: TodoPriority
  description?: string | null
  dueDate?: string | null
  tags?: string[]
}

export type UpdateTodoInput = {
  id: string
  title: string
  description: string | null
  dueDate: string | null
  priority: TodoPriority
  tags: string[]
  status: TodoStatus
}

export type MoveTodoInput = {
  todoId: string
  fromStatus: TodoStatus
  toStatus: TodoStatus
  toIndex: number
}

export const BOARD_COLUMNS: Array<{ status: TodoStatus; label: string }> = [
  { status: 'todo', label: 'To Do' },
  { status: 'in_progress', label: 'In Progress' },
  { status: 'done', label: 'Done' },
]

File: backend/lib/todo_api/todos/todo.ex
defmodule TodoApi.Todos.Todo do
  @moduledoc """
  Domain entity for Todo items.
  """

  @type priority :: :low | :medium | :high

  @type t :: %__MODULE__{
          id: String.t() | nil,
          title: String.t(),
          description: String.t() | nil,
          done: boolean(),
          due_date: DateTime.t() | nil,
          priority: priority(),
          tags: [String.t()],
          inserted_at: DateTime.t(),
          updated_at: DateTime.t()
        }

  @enforce_keys [:title, :done, :priority, :tags, :inserted_at, :updated_at]
  defstruct [
    :id,
    :title,
    :description,
    :due_date,
    :inserted_at,
    :updated_at,
    done: false,
    priority: :medium,
    tags: []
  ]
end
